### PR TITLE
Group Project : Recipe app

### PR DESCRIPTION
**Pull request by SanaTameem**
# Description :
The Recipe app keeps track of all your recipes, ingredients, and inventory. It will allow you to save ingredients, keep track of what you have, create recipes, and generate a shopping list based on what you have and what you are missing from a recipe. Also, since sharing recipes is an important part of cooking the app should allow you to make them public so anyone can access them.

- The app will have some common interfaces. These are some of the interfaces and some wireframes for some of them:
  - A login page.
  - A registration page.
  - A food list (with all CRUD implementation, except for 'update').
  - A recipes list (with all CRUD implementation, except for 'update').
  - A general shopping list view (all missing food for all your recipes and total price).
  - A list of all public recipes from other users with their names and total prices.

## Project requirements
- You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
- Login page and registration page:
  - Should be built with Devise.
- Food list (only for a group that has less than 3 members):
  - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
  - Should lead to a form that allows users to add new food.
- Recipes list:
Should display a list of recipes created by the logged-in user as in the wireframe.
  - Should lead to recipe details.
  - If the user is the owner of the recipe, should allow the user to delete it.
- Public recipe list:
  - Should display a list of all public recipes ordered by newest as in the wireframe.
  - Should lead to recipe details.
- Recipe details:
  - Should display a toggle button that allows for a recipe to be made public or private.
  - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
  - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
- General shopping list view (only for a group that has less than 3 members):
  - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
  - Should count the total food items and total price of the missing food.
- Make sure there are no N+1 queries happening.
- Create a navigation menu that allows users to open all of the pages you created.
- Write unit and integration tests

- Technical setup
  - Set up the repository on GitHub and use Gitflow.
  - Set up Devise for authentication.
  - Set up RSpec and Capybara testing libraries.

### CODE REVIEW

File: app/views/foods/_food.html.erb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<tr>
+<td class="t-item"><%= food.name %></td>
+<td class="t-item"><%= food.measurement_unit %></td>
+<td class="t-item">$<%= food.price %></td>
+<td class="t-item"><%= food.quantity %></td>
+<td>
+   <%= link_to "Delete", food_path(food), class: "delete",
```
Comment (TadesseDev) [inline app/views/foods/_food.html.erb:7]: - [Optional] Dear Team please consider updating the the validation for the price as it is possible to have a floating value as part of the price. Currently creating a price with a float number throws an error, which shouldn't be happening. 
### The issue 
Screenshot 2023-10-27 at 6 49 21 in the evening

File: app/views/recipes/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      </div>
+      <p>Cooking time: <%=@recipe.cooking_time%> minutes</p>
+   </div>
+   <p><%=@recipe.description%></p>
+   <div class='recipe_buttons'>
+   <%= button_to(@recipe.public ? "Make Private" : "Make Public", toggle_public_recipe_path(@recipe), method: :put, remote: true) %>
+      <% if can? :update, @recipe%>
+         <%=link_to "Generate Shopping List", general_shopping_list_index_path, class: "button" %>
+         <%=link_to "Add ingredient", new_recipe_recipe_food_path(@recipe), class: "button" %>
+      <%end%>
```
Comment (TadesseDev) [inline app/views/recipes/show.html.erb]: - [ ] Dear team you did a great job adding the public and private recipe, but the `make private` option should only be available to the owner of the recipe. Currently, I created the public recipe from user one and opened it from user two, I see an option to make the recipe private even though user two is not the owner of the recipe, also clicking the button ends up throwing an error. Please consider addressing this. 

### The issue 

https://github.com/SanaTameem/recipe-app/assets/69077061/047c8bb4-6966-4f02-8dcb-8cf365d0ad90

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi Team

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
212094426-b41dd457-c9db-4ab3-bbe9-176dd8dd08b7




## Highlights
 - [x] Good job overall following the template design ✅
 - [x] You have written a clean code ✅
 - [x]  Most of the requirement has been meet. ✅

 
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by gemmen29 — APPROVED** [review]
Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**